### DISCUSSION TITLE
Is there any way to get key code?

### DISCUSSION THREAD

**Discussion by teugenea** [emilk/egui#1590] [Q&A]
Hi! 
Egui uses virtual keyboard key codes to translate keyboard events to egui events. If a keyboard has multiple layouts I can check input for A-Z keys but not for '[' key for example.

Let's consider this code:
```rust
let events = ui.input().events.clone();
for event in &events {
    match event {
        egui::Event::Key{key, pressed, modifiers} => {
            println!("{:?} = {:?}", key, pressed);
        },
        egui::Event::Text(t) => { println!("Text = {:?}", t) }
        _ => {}
    }
}
```
It prints:
```
A = true
Text = "a"
A = false
```
I can check key 'A' by 'key' value.
But if I switch keyboard layout and press '[':
```
Text = "х"
```
Different output for different languages but I want my code to work the same way for any layout.

How can I get raw key code instead of egui::Event::Key?

**Comment by emilk** [comment]
I don't see it as egui:s job to communicate all input events (be it keyboard, joystick etc) to the user, but only those events that are relevant to a UI. I strain to think of a situation where a user interface (e.g. a web app) needs to be able to figure out if "the button left of the enter key" was pressed, no matter what keyboard layout was used. I also don't think the scancode is even available on web backends.

If you want low level control I suggest you use `winit` directly, like in https://github.com/emilk/egui/blob/master/egui_glow/examples/pure_glow.rs

**Reply by teugenea** [reply]
Ok, web backend may not support scancodes but what about virtual keyboard key codes?
And one more thing. Is Egui supposed to support hot keys? Without using virtual keyboard it will be hard to handle the same hot key on different layouts ( for example **Ctrl + ;** )

**Reply by dbalsom** [reply]
> I don't see it as egui:s job to communicate all input events (be it keyboard, joystick etc) to the user, but only those events that are relevant to a UI. I strain to think of a situation where a user interface (e.g. a web app) needs to be able to figure out if "the button left of the enter key" was pressed, no matter what keyboard layout was used. I also don't think the scancode is even available on web backends.

There are several emulators in development using egui & eframe - I'm writing an IBM PC emulator, my friend is writing a Macintosh emulator.  There's not much we can do about the web, but what I do is synthesize the hardware scancodes from winit KeyCodes.  There are tons of nitpicky details in converting modern keyboard layouts into their 1980's counterparts, and having every possible KeyCode available can mean the difference in a user being able to type a character in their native code page or not. 

Here's one workaround we came up with, which involves installing a hook to receive the raw winit events, but it's a bit of a hack:

https://github.com/twvd/egui/commit/856d675c30693d6650a417b1d8443b39d1636171

It would be really great if we could get access to the raw Winit event without needing to fork egui.   One of the main reasons I switched from doing winit + wgpu + egui manually, was the constant dependency treadmill keeping all the pieces working together, something that eframe solves nicely out of the box.  I pretty much gave up when winit hit 0.30, I don't think I need to tell you why.

Even a set of untranslated keycodes would work if the winit idea is unfeasible. Consider that someone may be using a US keyboard to emulate a Swedish keyboard layout.  If there are extra layers of translation here the whole thing becomes impossible to untangle.
